src/Servers/Kestrel/Core/src/Internal/Http3/Http3Connection.cs

```diff
@@ -288,54 +324,87 @@ private void UpdateStreamTimeouts(DateTimeOffset now)
                     Debug.Assert(streamDirectionFeature != null);
                     Debug.Assert(streamIdFeature != null);
 
+                    var context = CreateHttpStreamContext(streamContext);
+                    var pendingStream = new Http3PendingStream(context, streamIdFeature.StreamId);
```

Comment:
Isn't this being allocated for every request? This is a hot path where we want to avoid allocating and work in general, unless necessary.

What's the goal of pending/unidentified streams?

Has this been profiled? There shouldn't be extra cost from web transport on regular HTTP/3.

Reply:
I agree. I am not sure if there is a better way to do it but the main idea here is that we now need to read the first couple bytes from the stream before creating a stream object for it. This is because we now can't just assume that unidirectional streams are control streams (as webtransport has unidirectional streams). However, if we get a completely empty stream, the read will wait until it gets some data. We need a way to time it out and keep track of it. For this purpose, I created this `Http3PendingStream` concept that wraps unidentified streams and provides an abort and ticks counter for timeout reasons.

This was the approach that Chris and I came up with. Do you know of some better way?

In terms of profiling, I did this: https://github.com/dotnet/aspnetcore/pull/42097#issuecomment-1182585040.  It isn't a lot but, from what I understood, it isn't too much of a difference. Although, more profiling needs to be done before we can say it more definitively.

Reply:
Would making it a struct help fix this issue?

Reply:
At least we may consider pooling and re-using these pending streams. I don't think that's an issue that should block this PR, but add it to the follow-up list.

A struct doesn't help because it will get boxed when stored in the collection.

Reply:
Sounds good. I will add this to my issue about follow up PRs. Perhaps the best solution would actually be to rewrite streams entirely so that they follow a decorator pattern or something similar. That way we can add functionality as we identify the stream rather than having to create new objects. Resolving here.